Move to lucene BoostQuery

javanna opened the pull request:
Latest version of lucene deprecated Query#setBoost and Query#getBoost which made queries effectively immutable. Those methods need to be replaced with `BoostQuery` that wraps any query that needs boosting.

This commit replaces usages of setBoost with BoostQuery and adds it to forbidden-apis for prod code.

Usages of `getBoost` are only partially removed, as some will have to stay for backwards compatibility.

nik9000 commented on core/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java:
```diff
@@ -363,15 +352,15 @@ public static BlendedTermQuery dismaxBlendedQuery(Term[] terms, final float[] bo
-                    TermQuery termQuery = new TermQuery(terms[i], ctx[i]);
-                    if (boosts != null) {
-                        termQuery.setBoost(boosts[i]);
+                    Query query = new TermQuery(terms[i], ctx[i]);
```
I think you might be able to pull this out into a helper method.

javanna replied:
yes maybe, but it would look a bit messy cause the helper method would need too many arguments e.g. current_term, current_context, boosts. not sure it buys us much.

nik9000 commented on core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java:
```diff
@@ -23,13 +23,7 @@
-import org.apache.lucene.search.MultiPhraseQuery;
-import org.apache.lucene.search.PhraseQuery;
-import org.apache.lucene.search.Query;
+import org.apache.lucene.search.*;
```
Ah! The star imports come back. Its fun watching these things wash in and out like the tide.

javanna replied:
yea I dunno what I should do to be honest here. As our contributing guide says I don't fight my IDE, unless we go for a strict policy on killing star imports (might have missed that)

nik9000 replied:
Nah - you are right. Don't fight the IDE - without a strict policy its just silly. And we'd need a nice IDE configuration to make them never happen.

nik9000 commented on core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java:
```diff
@@ -779,7 +757,10 @@ private Query applySlop(Query q, int slop) {
             }
             pq = builder.build();
-            pq.setBoost(q.getBoost());
+            //nocommit why do we deal with boost here? isn't that what applyBoost does?
```
Yeah! Good point. Maybe applySlop is called after applyBoost? That isn't a good reason, but it might be the reason.

javanna replied:
it used to be, not anymore, so maybe that can change too...

jpountz replied:
This method takes a phrase query and is supposed to create an equivalent phrase query that just has a different value of the slop, so we need to transfer the boost?

Maybe the method should look like this now:

```java
    private Query applySlop(Query q, int slop) {
        float boost = 1f;
        Query underlyingQuery = q;
        while (underlyingQuery instanceof BoostQuery) {
            BoostQuery bq = (BoostQuery) underlyingQuery;
            boost *= bq.getBoost();
            underlyingQuery = bq.getQuery();
        }
        if (underlyingQuery instanceof PhraseQuery) {
            PhraseQuery pq = (PhraseQuery) underlyingQuery;
            PhraseQuery.Builder builder = new PhraseQuery.Builder();
            builder.setSlop(slop);
            final Term[] terms = pq.getTerms();
            final int[] positions = pq.getPositions();
            for (int i = 0; i < terms.length; ++i) {
                builder.add(terms[i], positions[i]);
            }
            pq = builder.build();
            pq.setBoost(boost);
            return pq;
        } else if (underlyingQuery instanceof MultiPhraseQuery) {
            ((MultiPhraseQuery) underlyingQuery).setSlop(slop);
            return q;
        } else {
            return q;
        }
    }
```

javanna replied:
are you sure that this complication is needed? It seems like this boost business was only needed because applySlop used to be called after applyBoost. Now that we call applyBoost afterwards, we should be ok? I might be missing something though :)

jpountz replied:
you may be right, but it sounds a bit dangerous to me to rely on the order in which methods are called?

javanna replied:
should we add a couple of asserts to verify that the boost never gets set before calling applyBoost, rather than handling a situation that never happens?

jpountz replied:
if we can assert that, it would work for me too

javanna replied:
I added two asserts, I haven't made up my mind on whether I prefer the asserts or your code snippets though :)  let me know what you prefer

nik9000 commented on core/src/main/java/org/apache/lucene/search/vectorhighlight/CustomFieldQuery.java:
```diff
@@ -103,8 +99,11 @@ private void convertMultiPhraseQuery(int currentPos, int[] termsIdx, MultiPhrase
-            PhraseQuery query = queryBuilder.build();
-            query.setBoost(orig.getBoost());
+            Query query = queryBuilder.build();
+            //nocommit not sure why we deal with boost here, given that we pass along the boost to flatten method too...
```
Errr. Oh man. I should know this one but I don't!

javanna replied:
indeed, you should know! :)

nik9000 replied:
OK - I think we are double counting the boost here. Its hard to tell though.

javanna replied:
I tend to agree, the fact that tests are green gives us a false yet comfortable sense of security :) If it wasn't highlighting stuff I would really look into writing the test that is missing, but...  I have no clue where to start

nik9000 replied:
There's just `HighlighterSearchIT` IIRC.

nik9000 commented on core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java:
```diff
@@ -64,8 +64,9 @@ public AllTermQuery(Term term) {
     public Query rewrite(IndexReader reader) throws IOException {
-        if (getBoost() != 1f) {
-            return super.rewrite(reader);
+        Query rewritten = super.rewrite(reader);
```
Does this change come from the boost change or something else? It feels related but not the same.

javanna replied:
it still does the same as before, but without using getBoost which is deprecated :) super.rewrite will return this if boost == 1f....

nik9000 replied:
Got it!

nik9000 commented on core/src/main/java/org/elasticsearch/common/lucene/search/Queries.java:
```diff
@@ -127,10 +127,13 @@ public static BooleanQuery applyMinimumShouldMatch(BooleanQuery query, @Nullable
+                return new BoostQuery(bq, query.getBoost());
+            }
+            return bq;
+        } else {
```
I usually prefer the way without the else but I don't object to either one.

javanna replied:
me too, but I find that this one is more readable this way...e.g. I forgot the second return above while changing this and introduced tons of failures :)

nik9000 replied:
Cool.

nik9000 commented on core/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java:
```diff
@@ -62,11 +63,21 @@ protected void doXContent(XContentBuilder builder, Params params)
-        return new SpanMultiTermQueryWrapper<>((MultiTermQuery) subQuery);
+        Query wrapper = new SpanMultiTermQueryWrapper<>((MultiTermQuery) subQuery);
+        if (boost != 1f) {
+            wrapper = new BoostQuery(wrapper, boost);
```
Should this be a `SpanBoostQuery`?

javanna replied:
good point...maybe... need to check why there are no failing tests around this, there should be. I think this is definitely a bug

jpountz replied:
+1 to SpanBoostQuery. I suspect it never fails because we have no test that uses a span-multi query as an inner query (as opposed to top-level)?

javanna replied:
moved to SpanBoostQuery and added a test that fails without it

nik9000 commented on core/src/test/java/org/elasticsearch/common/lucene/search/MultiPhrasePrefixQueryTests.java:
```diff
@@ -76,8 +78,11 @@ public void testBoost() throws Exception {
         IndexReader reader = DirectoryReader.open(writer, true);
         MultiPhrasePrefixQuery multiPhrasePrefixQuery = new MultiPhrasePrefixQuery();
         multiPhrasePrefixQuery.add(new Term[]{new Term("field", "aaa"), new Term("field", "bb")});
+        //nocommit: seems like this test still makes sense? or shall we rather forbid usage of setBoost already?
```
Can we forbid it in src/main but not src/test?

jpountz replied:
the test makes sense indeed. +1 to ban Query.setBoost in src/main but not src/test like @nik9000 suggests. I think this should work if you only touch dev-tools/src/main/resources/forbidden/core-signatures.txt instead of all-signatures.txt.

javanna replied:
I banned setBoost

nik9000 commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -424,26 +415,45 @@ protected final void assertLuceneQuery(QB queryBuilder, Query query, QueryShardC
     protected abstract void doAssertLuceneQuery(QB queryBuilder, Query query, QueryShardContext context) throws IOException;
 
+    protected static void assertTermOrBoostQuery(Query query, String field, String value, float fieldBoost) {
+        if (fieldBoost != AbstractQueryBuilder.DEFAULT_BOOST) {
```
And if it is DEFAULT_BOOST can you assert the opposite?

javanna replied:
don't I do it below by calling assertTermQuery?

nik9000 replied:
Ah! Got it.

jpountz commented on core/src/main/java/org/elasticsearch/index/query/HasChildQueryBuilder.java:
```diff
@@ -215,8 +215,9 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
-
+        //nocommit: this should be ok to drop given that the parent class will wrap the result
+        // of this method call into BoostQuery when needed? Otherwise we would apply boost twice?
+        //innerQuery.setBoost(boost);
```
+1

jpountz commented on core/src/main/java/org/elasticsearch/index/query/HasParentQueryBuilder.java:
```diff
@@ -129,7 +129,9 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
-        innerQuery.setBoost(boost);
+        //nocommit: this should be ok to drop given that the parent class will wrap the result
+        // of this method call into BoostQuery when needed? Otherwise we would apply boost twice?
+        //innerQuery.setBoost(boost);
```
+1

jpountz commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -424,26 +415,45 @@ protected final void assertLuceneQuery(QB queryBuilder, Query query, QueryShardC
+                    assertThat(boostQuery.getBoost(), equalTo(queryBuilder.boost()));
+                    query = boostQuery.getQuery();
+                }
+            }
```
can you assert that the boost is the default boost at this point?

jpountz replied:
nevermind, just saw the comments below

jpountz commented:
Thanks for doing this work! I left some comments but things look good to me in general. As highlighted in ones of the comments, I think we should forbid usage of Query.setBoost to prevent it from coming back.

javanna commented:
@jpountz pushed a new commit, can you have another look please?

javanna commented on core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java:
```diff
@@ -779,7 +757,9 @@ private Query applySlop(Query q, int slop) {
-            pq.setBoost(q.getBoost());
+            //make sure that the boost hasn't been set beforehand, otherwise we'd lose it
+            assert q.getBoost() == 1f;
+            assert q instanceof BoostQuery == false;
```
@jpountz lemme know what you think about this. I wouldn't want to handle a case that's not supported, on the other hand I am not loving these asserts either.....

jpountz replied:
that works for me, let's just make sure that we have a unit test that sets both the slop and the boost on a phrase query

javanna replied:
done

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java:
```diff
@@ -722,16 +720,25 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
+            query = new BoostQuery(query, boosts.get(i));
+        }
+
+        return query;
```
I think that this unwrapping business is needed only in query_string as it's the only query that supports providing boosts as part of the query, thus the resulting boolean query might have been boosted multiple times. I think the same cannot happen with simple query string , multi match query and similar queries that call Queries#applyMinimumShouldMatch... but I might be missing something

javanna commented on core/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java:
```diff
@@ -62,11 +65,26 @@ protected void doXContent(XContentBuilder builder, Params params)
+            subQuery = boostQuery.getQuery();
+            boost = boostQuery.getBoost();
+        } else if (subQuery instanceof SpanBoostQuery) {
+            //this should never happen as MultiTermQuery extends SpanQuery, but who knows...
```
I added this if branch. Don't think it ever happens and in this case it didn't seem a huge issue to support it rather than asserting against it.

jpountz replied:
hmm, MultiTermQuery does not extend SpanQuery, does it?

jpountz replied:
maybe you meant "does NOT"?

javanna replied:
yea I meant as none of the multi term queries in lucene extend SpanQuery . Shall we simply drop this branch then?

jpountz replied:
let's have an assert

javanna replied:
done

javanna commented:
I rebased and left a few inline comments for @jpountz , I think it's getting closer. Also removed the WIP label, this is not work in progress anymore.

jpountz commented on core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java:
```diff
@@ -722,16 +720,25 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
+            BoostQuery boostQuery = (BoostQuery) query;
+            boosts.add(boostQuery.getBoost());
+            query = boostQuery.getQuery();
+        }
```
no need to remember all individual boosts, you can just multiply:

```java
float boost  = 1f;
while(query instanceof BoostQuery) {
  BoostQuery boostQuery = (BoostQuery) query;
  boost *= boostQuery.getBoost();
  query = boostQuery.getQuery();
}

```

javanna replied:
I dunno, I thought it's better not to replicate the internal lucene behaviour? (the fact that it multiplies boosts).

jpountz replied:
I'm ok either way, just proposed this one because it looks simpler to me

javanna commented:
@jpountz I addressed your comments, left the last question about manually multiplying boosts vs letting lucene do it and maintain the original wrapped structure. Thanks for your time!

jpountz commented:
LGTM
